conjure_oxide/src/unstable/solver_interface/adaptors/minion.rs

```diff
@@ -0,0 +1,306 @@
+use std::collections::HashMap;
+use std::sync::{Condvar, Mutex, OnceLock};
+
+use crate::unstable::solver_interface::SolverMutCallback;
+use crate::unstable::solver_interface::{states, SolverCallback};
+
+use super::super::model_modifier::NotModifiable;
+use super::super::private;
+use super::super::stats::NoStats;
+use super::super::SolveSuccess;
+use super::super::SolverAdaptor;
+use super::super::SolverError;
+use super::super::SolverError::*;
+
+use minion_rs::run_minion;
+
+use crate::ast as conjureast;
+use minion_rs::ast as minionast;
+
+/// A [SolverAdaptor] for interacting with Minion.
+///
+/// This adaptor uses the `minion_rs` crate to talk to Minion over FFI.
+pub struct Minion;
+
+static MINION_LOCKED: Mutex<bool> = Mutex::new(false);
+static MINION_CONDVAR: Condvar = Condvar::new();
+static USER_CALLBACK: OnceLock<Mutex<SolverCallback>> = OnceLock::new();
+
+fn minion_rs_callback(solutions: HashMap<minionast::VarName, minionast::Constant>) -> bool {
+    #[allow(clippy::unwrap_used)]
+    let callback = USER_CALLBACK
+        .get_or_init(|| Mutex::new(Box::new(|x| true)))
+        .lock()
+        .unwrap();
+
+    let mut conjure_solutions: HashMap<conjureast::Name, conjureast::Constant> = HashMap::new();
+    for (minion_name, minion_const) in solutions.into_iter() {
+        let conjure_const = match minion_const {
+            minionast::Constant::Bool(x) => conjureast::Constant::Bool(x),
+            minionast::Constant::Integer(x) => conjureast::Constant::Int(x),
+            _ => todo!(),
+        };
+
+        //FIXME (niklasdewally): what about machine names?
+        let conjure_name = conjureast::Name::UserName(minion_name);
+        conjure_solutions.insert(conjure_name, conjure_const);
+    }
+
+    (**callback)(conjure_solutions)
+}
+
+impl private::Sealed for Minion {}
+impl SolverAdaptor for Minion {
+    type Model = minionast::Model;
+    type Solution = minionast::Constant;
+    type Modifier = NotModifiable;
+
+    #[allow(clippy::unwrap_used)]
+    fn solve(
+        &mut self,
+        model: Self::Model,
+        callback: SolverCallback,
+        _: private::Internal,
+    ) -> Result<SolveSuccess, SolverError> {
+        // our minion callback is global state, so single threading the adaptor as a whole is
+        // probably a good move...
+        #[allow(clippy::unwrap_used)]
+        let mut locked = MINION_LOCKED.lock().unwrap();
```

Comment:
Why can't the locking just be a Mutex<()>, you lock it here, keep it until the end of this function, and when you drop it Rust will automatically give it to someone else (so no need for the notify?)

Reply:
I've never been fully sure about the difference between convar and mutexes - I agree that a mutex would probably be good enough here.

I guess the performance impact of condvars.comes into play when we want to lock depending on the value inside the mutex? This would avoid busy waits like while (a.lock() == false) {}? I don't need a value, just mutual exclusion, so this is irrelevant...

Thank you for clarifying this for me :)

Reply:
Generally, a convar is for when you need to wakeup / signal another thread (for example, there is some work todo), here you just need mutual exclusion, so a Mutex should be fine by itself (you could actually put some data in the mutex, usually in Rust we do, but something you find yourself protecting an external resource, and then you just use Mutex<()>).